### ISSUE TITLE
Apple silicon/ Blender 3.4.0 / MolNodes 2.0.2 package installation failure

### ISSUE THREAD

**Issue by PlethoraChutney** [BradyAJohnston/MolecularNodes#108]
Hi Brady! Excited to play with the new features, and maybe finally dip my toes into contributing!

Unfortunately, trying to install the required packages using this button in blender:

<img width="248" alt="image" src="https://user-images.githubusercontent.com/27385979/207387948-e2f0a682-32b7-4292-9241-c5069a5903fb.png">

results in an error message reading:

> Failed to install required packages. Please check /Users/posert/Library/Application Support/Blender/3.4/scripts/addons/MolecularNodes/logs/side-packages-install.log

The contents of that log file are:

```
$ cat ~/Library/Application\ Support/Blender/3.4/scripts/addons/MolecularNodes/logs/side-packages-install.log
-----------------------------------
Installer Started: 2022-12-13 08:18:36.297157
-----------------------------------
###################################
Installer finished: 2022-12-13 08:19:05.242502
###################################
```

# System Info
* M2 macbook air
* Monterey 12.6.1
* Blender 3.4.0
* MolNodes 2.0.2

# Steps already taken
* Uninstall molnodes, restart blender, re-install molnodes

**Comment by BradyAJohnston** [comment]
How incredibly unhelpful of a log file haha. (also envoius of the M2). 

Can you try launching Blender via the terminal `blender` or `/Applications/Blender` should launch it. Then when you try to install the packages again it should hopefully give a more helpful error message printed to the terminal.

**Comment by PlethoraChutney** [comment]
Let me tell ya, if I'd known how nice these machines are I would have "accidentally" destroyed my old Dell ages ago.

Ah, good call. It seems Blender is failing to build a wheel for `numpy` during the `biotite` install. I'll poke around a bit! Thanks!

**Comment by BradyAJohnston** [comment]
Are you getting the same compiler error as #102 where it's lacking `gcc`?

**Comment by PlethoraChutney** [comment]
No, I'm missing `python.h`. Trying to figure out why, since I should have it. I'm now just failing to install MDAnalysis, I got numpy working.

**Comment by BradyAJohnston** [comment]
Okay this is a good test case. This seemed to be an issue with earlier versions of Blender when I was originally developing the addon which is why I had the complicated installation od MDAnalsysis in an external python and you had to then link it to Blender's python. 

It might be that it's the ARM build of Blender doesn't include the `python.h` file but it's now included in other versions?

**Comment by PlethoraChutney** [comment]
That was my thought as well, but I'm also failing to install MDAnalysis in my machine's own python, as well as conda! So I don't think I can blame Blender unfortunately haha. I'm getting close to just nuking all the pythons on my machine and starting over

**Comment by BradyAJohnston** [comment]
There comes a time in every computers life when one must remove any trace of the word python and start again

**Comment by PlethoraChutney** [comment]
*taps plays*

**Comment by BradyAJohnston** [comment]
🫡

**Comment by Luc1100** [comment]
I'm on an M1 Pro, Ventura 13.1, Blender 3.4.0, MolNodes 2.0.2 and I am having the same issue `SystemError: Cannot compile 'Python.h'. Perhaps you need to install python-dev|python-devel.` However, I'm not having any issues downloading MDAnalysis on my machine's own python.

**Comment by BradyAJohnston** [comment]
You could likely solve the issue immediately with the solutions in #24 but I'm more interested as to why it installs OK on windows / intel macs but not on ARM macs. Potentially a Blender distro issue

**Comment by Luc1100** [comment]
Soft linking a conda environment with the packages installed worked, hopefully you can get the one click install working soon!

**Comment by BradyAJohnston** [comment]
If it's an issue with the ARM compiled Blender then I may have to include a second-step of the install only for M1/M2 macs which copies how it was achieved in the previous version. It's not ideal as it significantly complicates the installation process unless somebody can figure something else out. 

@Luc1100 can you share the exact steps that you followed to get it working? Does it continue to work after you restart Blender? Are you able to use the Molecular Dynamics import panel or just the PDB / Local File import?

**Comment by Luc1100** [comment]
To get it working I made a new conda environment
`conda create --name blender python==3.10 biotite==0.35.0 mdanalysis==2.2.0 -c conda-forge -y`

Found the path to the environment
`/path/to/miniconda3/envs/blender`

In `/Applications/Blender.app/Contents/Resources/3.4`, I renamed the python that Blender came with
`mv python _python`

Then I made a soft-link to the anaconda version of python
`ln -s /path/to/miniconda3/envs/blender python`

It has worked after every blender restart and the molecular dynamics import panel is there. The only issue that I am running into is probably user error (I import a trajectory and I can only see the points in wireframe mode regardless of the render engine)

**Comment by PlethoraChutney** [comment]
This works for me as well!

I am still but a lowly pip user --- using conda worked right away. I noticed that Blender is also using pip. Perhaps that's the issue.

Maybe this will finally get me to learn what the hell conda is.

**Comment by maxewilkinson** [comment]
I'm having the same problem, but @Luc1100 fix of linking a conda environment doesn't work for me. When I do this, Blender crashes upon startup. Hmm any ideas for other work arounds? 

Using MacBook Pro with M1 Max, Monterey 12.5, Blender 3.4.1, MolNodes 2.0.2

**Comment by maxewilkinson** [comment]
It looks like my Blender was crashing because the conda environment was using x86_64 (via Rosetta?) instead of arm64. So Blender was trying to call arm-compiled python packages and getting x86 packages instead. I don't know if this is just a peculiarity of how I installed conda.

Anyway, to fix this, when you create the conda environment, use:

`CONDA_SUBDIR=osx-arm64 conda create -n molnodes python==3.10 biotite==0.35.0 mdanalysis==2.2.0 -c conda-forge -y`
`cd /Applications/Blender-3.4.app/Contents/Resources/3.4`
`mv python old_python`
`ln -s /Users/maxwilkinson/opt/anaconda3/envs/molnodes python`

Molecular Nodes now installs, but failed to import structures, because conda installed numpy 1.24.0, in which `np.int` used by Molecular Nodes is deprecated. This is then fixed by

`conda activate molnodes`
`CONDA_SUBDIR=osx-arm64 conda install numpy=1.22`

So far this is working for me!

**Comment by BradyAJohnston** [comment]
Hi @maxewilkinson thanks for the bug report! Apple M1 / M2 continues to be a source of trouble, but I appreciate the troubleshooting you did and the fix that you posted. 

I'll have a look into the `np.int` as that will be problematic.

**Comment by Aseamann** [comment]
Max's comment is what worked for me. I believe it ensures what gcc complier is being used?

https://github.com/BradyAJohnston/MolecularNodes/issues/108#issuecomment-1365540371

**Comment by BradyAJohnston** [comment]
Thanks @Aseamann for the details. Might require some more specifics for installation on M1 macs. Unfortunately I don't have a machine to test on, and I'm not sure what the approach would be to try and help with this from the package perspective.

**Comment by PlethoraChutney** [comment]
Here's a gist for the fix, which might come in handy if you're giving a workshop on MolNodes

https://gist.github.com/PlethoraChutney/fd34e3137044cb1e1efc881157e0e17b

**Comment by BradyAJohnston** [comment]
many thanks! I have a workshop coming up in a week and this will likely come in handy

**Comment by YaoYinYing** [comment]
Here is an optional:

1. Create a new conda env containing python3.10: `conda create -n py310 python=3.10 -y`
2. Activate it: `conda activate py310`
3. walk to the build-in python binary dir of blender: `cd /Applications/Blender.app/Contents/Resources/3.4/python/bin`
4. run pip with a CPATH anoucement: `CPATH=$(dirname $(which python))/../include/:$(dirname $(which python))/../include/python3.10:$CPATH ./pip3.10 install -r /Users/yyy/Documents/protein_design/MolecularNodes/MolecularNodes/requirements.txt` (I try to put the required packages in a `requirements.txt`, but one can replace it with the `MDAnalysis==2.2.0`)

**Comment by caitlinmonney** [comment]
Sadly, I am in the same situation and I have not yet found a solution. Sam install package button with same error message. 
Have tried to install miniconda, no work
Using Mac Studio Apple M1 Max , Blender 3.4.1 , MolNodes 2.3.1
Please help

**Comment by BradyAJohnston** [comment]
Hi @caitlinmonney, my recommendation would be to completely uninstall the Molecular Nodes addon and completely uninstall Blender.

Once you have a clean system, reinstall Blender. **Before** you install Molecular Nodes, use the [script](https://gist.github.com/PlethoraChutney/fd34e3137044cb1e1efc881157e0e17b) posted by @PlethoraChutney to build the required python libraries. Once that is complete, then try installing Molecular Nodes and see how that goes.

**Comment by PlethoraChutney** [comment]
Unfortunately someone I was helping who had an M1 chip was not able to resolve the issue --- the packages failed to install using Blender, and the workaround caused a crash-on-launch. We still haven't been able to figure out a fix (and I'll admit, I'm a bit confused how something that works on an M2 machine fails on an M1 machine).

**Comment by YaoYinYing** [comment]
Hi @caitlinmonney I got a slighter solution of more elegance at https://github.com/BradyAJohnston/MolecularNodes/issues/108#issuecomment-1429384983 by borrowing the Python.h from conda installed `python3.10`. Would you mind taking a try on that ? 

And @BradyAJohnston I think this might work for most common cases.

**Comment by BradyAJohnston** [comment]
@PlethoraChutney I think I helped someone with an issue where Blender would crash on startup. Try reinstalling Blender, MN & the conda environment which seemed to fix the issue for them.

**Comment by BradyAJohnston** [comment]
@YaoYinYing would there be a way we could generalise this / make it simpler for users? If it works reliably can we enable this better on the MN installation side of things?

**Comment by BradyAJohnston** [comment]
I don't actually know much about C header files, but would we be able to just ship MN with our own `python.h` file? 

I've had a look at the `python.h` that was included with my Intel Mac Blender, could we just include some version of this? https://gist.github.com/BradyAJohnston/09737a5eb30e30f79e46795cb942ab7d

**Comment by YaoYinYing** [comment]
@BradyAJohnston Well, I don't know much about the C header either. It looks like the build-in python has truncated quite a few things.  I am working in progress on this automatical script based on the following steps:

1. determine whether the platform is macos on ARM;
2. check if conda or homebrew exists  ; 
3. check if python3.10 on ARM is installed via conda or homebrew; 
4. if the desired python3.10 does not exists, run conda install or brew install (if brew install fails try reinstall from source building); 
5. call `python3.10-config --include` and replace `?-I` with ':' and store the result as `HEADER_INCLUDE_DIR`; 
6. run 'CPATH=$HEADER_INCLUDE_DIR:$CPATH install -r requirements.txt'

Here is a phenotype script generated by ChatGPT: https://gist.github.com/YaoYinYing/8d6c297c0af30087bbefd0b2e73e47c9
I haven't carefully checked it but this idea might be feasible.

**Comment by YaoYinYing** [comment]
Also , the guideline at https://github.com/BradyAJohnston/MolecularNodes/issues/108#issuecomment-1429384983 has been updated to support brewed python3.10. 

See logs at https://gist.github.com/YaoYinYing/36b793fbb751fbceb320ee9437a6d77b.

**Comment by YaoYinYing** [comment]
I'm not sure how the `python.h` at `./Contents/Resources/3.4/python/lib/python3.10/site-packages/debugpy/_vendored/pydevd/pydevd_attach_to_process/common` works for Blender. In my mind, C & Python are case-sensitive. So I guess the 'Python.h not found' will still occur if not external CPATH is anounced before `pip install`.

**Comment by BradyAJohnston** [comment]
We could also switch back to creating the conda environment, then just adding that to Blender's python `$PATH` which was a little finicky, at least it didn't crash all of Blender when it didn't work.

**Comment by YaoYinYing** [comment]
@BradyAJohnston I think i have solved this one-click installation via conda environment. Feel so excited! https://gist.github.com/YaoYinYing/00f86f2cc400c7e9db335f6211d512f2

If you too think this solution is okay I will create a new PR for this fix!

**Comment by caitlinmonney** [comment]
First of all, thank you all very very much for being so proactive and trying to help out. I really appreciate it. 

OK, so I have tried uninstalling and reinstalling everything in the suggested sequence. Still no success, I am not given the option to install atomise from the Add-ons install menu-as seen in our YT video-, I do get the Install Packages button and the error message as mentioned above.  

With the help of @marwan-abdellah we have tried     
pip install biotite==0.35.0 MDAnalysis==2.2.0   
But no success
Tried to uninstall the default numpy and then let the deps pull the appropriate version
But still no success

I have tried the latest MN version with the latest Blender release on my MacBook Air Intel i5, all seems to work. Making my was through your tutorials to see if anything comes up. But I would very much like to find a fix for my M1 as this is my main machine.  We-re going to continue to trouble/shoot on our end and cross our toes we find a fix. Let us know if you have further suggestions...

Thanks

**Comment by YaoYinYing** [comment]
Finally this happens to me again - I have no idea why it just works on my machine lol

Maybe I should visit to the apple store nearby and look for the ARM Macs for more tests ....

**Comment by YaoYinYing** [comment]
@caitlinmonney Hi I just wrote a script for this issue. Would you please help me for a test? It would be fast.

You should do the following:

1. Remove Blender.app
2. Remove the user data of Blender (it would be at "~/Library/Application Support/Blender"): `rm -r ~/Library/Application\ Support/Blender`
3. Reinstall the latest Blender
4. Launch Blender and install the latest MN
5. Wxecute `curl -fsSL https://raw.githubusercontent.com/YaoYinYing/MolecularNodes/ARM-mac-installer/MolecularNodes/troubleshoot/InstallDependencies_M1.sh | $SHELL` to help you install the dependencies. If it fails, just repeat this step once.
6. check the MN to see if you can load a PDB.

**Comment by BradyAJohnston** [comment]
> Maybe I should visit the apple store nearby and look for the ARM Macs for more tests ....

If somebody wants to bankroll an M1 / M2 mac for me to debug on, that would certainly be appreciated haha.

While this can potentially be a workaround for people, I don't like the idea of doing so much installation via the 'Install Packages' button etc. I'm chatting with some of the Blender devs about it. Hopefully I can get some more information about what the best way going forward might be.

**Comment by BradyAJohnston** [comment]
@caitlinmonney please do let us know any solutions that you might come up with. I'm pretty stuck with possible solutions. Without having access to an M1 / M2 machine to test & debug on I can't really be of much help unfortunately, and it's up to others with ARM machines to share possible solutions they have found that work for them.

**Comment by YaoYinYing** [comment]
> I don't like the idea of doing so much installation via the 'Install Packages' button etc.

Me too. This has done so much unnecessary stuff for this issue.
I will be very appreciated if someone in Blender dev team get python-dev inside this bundled Python exec.

**Comment by BradyAJohnston** [comment]
> I have tried the latest MN version with the latest Blender release on my MacBook Air Intel i5, all seems to work. Making my was through your tutorials to see if anything comes up. But I would very much like to find a fix for my M1 as this is my main machine. We-re going to continue to trouble/shoot on our end and cross our toes we find a fix. Let us know if you have further suggestions...

@caitlinmonney and others who are able to get it installing on another machine. As a workaround, once you have the molecules imported, you can then open the `.blend` files on your main machine. The python packages are only required for the initial import step. All of the nodes should still work on your M1 / M2 machines without the need for the external python packages at all, as they are only required for importing.

**Comment by BradyAJohnston** [comment]
I've been chatting with some Blender devs, and it seems that it's likely not a `Python.h` problem, as Blender shouldn't be shipping with the required files for compiling packages.

What is likely causing the discrepancy between systems, is the lack of `.whl` that is pre-built on PyPi for M1 / M2 macs. Looking at the MDAnalysis PyPi page, they are missing an `arm64` version which packages like SciPy have available. This causes the issue on M1 / M2 macs when they have to try and compile the packages themselves but don't have the pre-requisite files. This would also explain why earlier versions of Molecular Nodes had to handle package installation differently, when I was installing the most recent versions which may not have had the necessary `.whl` files and caused issues on more systems, whereas now it is limited to ARM macs.

## MDAnlysis PyPi v2.2.0
[![image](https://user-images.githubusercontent.com/36021261/223128492-f389b827-5c10-4662-81a3-ebefbf50f114.png)](https://pypi.org/project/MDAnalysis/2.2.0/#files)

## MDAnlysis PyPi v2.4.2
[![image](https://user-images.githubusercontent.com/36021261/223129127-7310c4b6-919b-498b-9fb5-6906868fedb0.png)](https://pypi.org/project/MDAnalysis/#files)

## SciPy PyPi
[![image](https://user-images.githubusercontent.com/36021261/223129840-c27dbb08-8080-46a3-a3a0-9fc61abee962.png)](https://pypi.org/project/scipy/#files)

# A solution?

I've chatted with the MDAnalysis devs on a few issues before, so I might be able to see if they can also release an ARM build of MDAnalysis which would potentially alleviate the installation issues. I don't really know much about python releases so I don't know if this is something that has to be done by them or can be built by PyPi somehow? Could we (somebody with an ARM mac) build a version or MDAnalysis for ARM macs, that we could include for installation on M1 / M2 macs which might solve the issues? 

Anyone with some more information or knowledge about the whole system works please do let me know.

**Comment by BradyAJohnston** [comment]
I've contacted the MDAnlaysis devs and they are going to try and set up MacOS ARM builds on PyPi for the `2.5.0` release which should hopefully resolve this issue!! (https://github.com/MDAnalysis/mdanalysis/issues/4054)

**Comment by YaoYinYing** [comment]
> Could we (somebody with an ARM mac) build a version or MDAnalysis for ARM macs,

Hi @BradyAJohnston I dig around the cache dir and find the built wheel of MDA and uploaded at here:
[MDAnalysis-2.2.0-cp310-cp310-macosx_11_0_arm64.whl.zip](https://github.com/BradyAJohnston/MolecularNodes/files/10904981/MDAnalysis-2.2.0-cp310-cp310-macosx_11_0_arm64.whl.zip)

The origin file for record sha256 stuff:
[origin.json.zip](https://github.com/BradyAJohnston/MolecularNodes/files/10905006/origin.json.zip)


Hope this may be helpful for you.

**Comment by YaoYinYing** [comment]
> I've contacted the MDAnlaysis devs and they are going to try and set up MacOS ARM builds on PyPi for the `2.5.0` release which should hopefully resolve this issue!! ([MDAnalysis/mdanalysis#4054](https://github.com/MDAnalysis/mdanalysis/issues/4054))

Wow I can't wait to see it! Then I can close that fix PR full of bullshit!

**Comment by YaoYinYing** [comment]
Looks like the lack of a target prebuild `wheel `file from the official repo may trigger `pip` to build it from scratch on local machine using `gcc/clang`(installed via `xcode-select`). 

What I guess is in this circonstance, `Python.h` will be required as it has been anounced in a specific `C/CPP` file of MDA source code...

**Comment by YaoYinYing** [comment]
Wheels can be build using 
```shell
# build wheel
CPATH=$CPATH:$(python3.10-config --include | sed  -E 's/ ?-I/:/g') ./pip3.10 wheel MDAnalysis==2.4.2 --cache-dir ../cache

# remove downloaded wheels
rm *.whl

# find built MDA wheel
find ../cache/wheels/ -name "*.whl" |grep MDA
```

**Comment by BradyAJohnston** [comment]
I still need an M1 to test. @PlethoraChutney would you be up for trying to build a wheel, then installing that in a vanilla blender install?

**Comment by PlethoraChutney** [comment]
For sure. I just gotta know where to run the above commands!

**Comment by YaoYinYing** [comment]
@PlethoraChutney Well if you already have a global Python3.10, whether it was installed via brew or conda, you may create any new directory and run `pip wheel` into it. 

The wheel file shall be downloaded and built.

I guess gcc is still required for running a wheel buiding from source code contains C files.

```shell
mkdir MN_wheels;
cd MN_wheels;
pip wheel MDAnalysis==2.2.0 --cache-dir .
```

Then all the wheels will be appeared.

**Comment by PlethoraChutney** [comment]
OK, got that installed. I also had to install cython or the build would fail, so here's the steps to recreate:

```
conda create -n wheel-builder python=3.10 cython
conda activate wheel-builder
python -m pip wheel MDAnalysis==2.2.0 --cache-dir .
```

Now, sorry to be a little computer baby, but how do I convince blender to use this wheel?

**Comment by YaoYinYing** [comment]
@PlethoraChutney Great, then you can walk down to the Blender-bundled-python directory, generally it would be like `/Applications/Blender.app/Contents/Resources/3.4/python/bin`, and run pip install `./python3.10 -m  pip install <wheel>` . 

I am not sure whether the `wheel` is required by this pip install or not.

**Comment by BradyAJohnston** [comment]
If this is works, we can maybe ship the wheel built on your ARM machine as an apple M1 / M2 specific install. At least until the next release of MDAnalysis which might also fix the issue with supplying wheels on PyPi.

**Comment by YaoYinYing** [comment]
Await is all we need. 🤣 It should be fine to just pin the version as 2.5.0 after it's released.

**Comment by evanbunnage** [comment]
After failing to get MolecularNodes installed on my M1 machine (symptom was that blender wouldn't open after symlinking a conda env with biotite & MDA) I came to this thread. 

Can confirm rebuilding the wheels worked for me! 

```
mkdir ~/biotite-wheel
conda create -n wheel-builder python=3.10 cython
conda activate wheel-builder
python -m pip wheel biotite==0.35.0 --cache-dir .
conda deactivate

cd /Applications/Blender.app/Contents/Resources/3.4/python/bin
./python3.10 -m pip install biotite --cache-dir ~/biotite-wheel
```

I did the same thing for MDA but I haven't tested it, but can confirm biotite works and I can import structures.

**Comment by evanbunnage** [comment]
I just tried the Covid spike protein example and I can confirm that MDA works using the same procedure.

**Comment by BradyAJohnston** [comment]
@evanbunnage thanks so much for the details! Exciting that this is working. This seems like it will be a more robust fix for the time being. I'll try and update the docs soon for installation instructions.

**Comment by YaoYinYing** [comment]
Great! I'm sure customized wheel building should work better bcs of higher compatibility with one's own machine. Before the 2.5.0 official wheel release of MDA, this can be a solid fix.
BTW, biotite has an official wheel at 0.36.0 so  we can set it in requirements.txt for the fix too.

**Comment by oscarbiesanz** [comment]
I was still having issues with Blender finding python.h after trying the current proposed solution, but I'm not at all knowledgeable about computer science, so I thought I would ask GPT4 to help so that I could provide more than a problem, and I actually found a working solution. Let me know if this works for anyone else still having issues.

Everything seemed to resolve on the attempt where GPT4 suggested this command `cd /Applications/Blender.app/Contents/Resources/3.4/python/include/python3.10
ln -s ../../Python.framework/Versions/3.10/Headers/ Python.h
` was proposed after building the wheels. 

It also took me an embarrassing amount of time to realize I had to add the --user option to the final install command.

Let me know if this helps anyone!

**Comment by danpf** [comment]
FWIW, I just cloned the master branch, zipped up the MolecularNodes folder, installed it, closed out of blender, opened blender, clicked install biotite, waited, closed out of blender, and opened it again and it appears to work fine.

**Comment by ErikPoppleton** [comment]
Just wanted to throw in my $0.02 on this.  The method from @evanbunnage which is also in the documentation didn't work for me because of the Python.h issue for MDAnalysis.  I resolved it by getting the header libraries from [the official Python page](https://www.python.org/downloads/source/) for the version of Python which ships with Blender (3.10.9, for Blender 3.5). Unfortunately, if Blender is installed in either `/Applications` or `$HOME/Applications/`, Apple's System Integrity Protection prevents you from copying the contents of the `Include` directory from the downloaded version of Python, so you have to copy the whole Blender directory somewhere else, copy the include directory, then move Blender back.

I can then install MDA from the cached wheel and import it from Blender's internal Python console, but Molecular Nodes cannot see MDA and says its not installed.

I'm on an M1 Mac running Blender 3.5.

**Comment by BradyAJohnston** [comment]
Hi @ErikPoppleton this is an interesting one. It might be that there is a version mismatch for MDAnalysis. If you can import it via the python console inside of Blender then it should be available to MolecularNodes. What version do you get with the below. MolecularNodes is currently expecting `2.2.0` exactly.
```python
MDA.__version__
```

**Comment by ErikPoppleton** [comment]
Yes! The version was 2.4.3.  Downloading and caching specifically 2.2.0 fixed the problem and Molecular Nodes now sees MDA as installed.  Thanks!

**Comment by BradyAJohnston** [comment]
MDAnalysis has just released `2.5.0` which includes pre-built `.whl` for M1 / M2 macs, which should HOPEFULLY fix the installation error that has been plaguing us for so long. 

I don't have an apple silicon machine to test. I have created a potential new release which requests the new MDAnalysis. Can people please try this out? If possible on a fresh installation of Blender. If not a fresh installation, please ensure you restart Blender after clicking 'install' button.

https://github.com/BradyAJohnston/MolecularNodes/releases/tag/v2.7.1_beta

**Comment by Luc1100** [comment]
@BradyAJohnston I tried the new MolecularNodes install with a fresh blender install. I get the following error when I try the MDAnalysis install.

```
bpy.ops.mol.install_package(package="MDAnalysis", version="2.5.0", description="Install required python package: MDAnalysis")
Error installing package. Please check the log files in '/Users/lucasmorley/Library/Application Support/Blender/3.5/scripts/addons/MolecularNodes/logs'.
```

Some of the stuff in side-packages-install.log is:
```
INFO:root:Running Pip: '['/Applications/Blender.app/Contents/Resources/3.5/python/bin/python3.10', '-m', 'pip', 'install', 'MDAnalysis==2.5.0']'
ERROR:root:Command failed: ['/Applications/Blender.app/Contents/Resources/3.5/python/bin/python3.10', '-m', 'pip', 'install', 'MDAnalysis==2.5.0']
ERROR:root:stdout: Collecting MDAnalysis==2.5.0
  Using cached MDAnalysis-2.5.0-cp310-cp310-macosx_11_0_arm64.whl (3.3 MB)
```
...
```
gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/brecht/dev/build_darwin/deps/Release/sqlite/include -I/Users/brecht/dev/build_darwin/deps/Release/bzip2/include -I/Users/brecht/dev/build_darwin/deps/Release/lzma/include -I/Users/brecht/dev/build_darwin/deps/Release/zlib/include -mmacosx-version-min=11.00 -arch arm64 -mmacosx-version-min=11.00 -arch arm64 -I/Applications/Blender.app/Contents/Resources/3.5/python/include/p
ython3.10 -c Bio/Align/_aligners.c -o build/temp.macosx-11.00-arm64-cpython-310/Bio/Align/_aligners.o
      Bio/Align/_aligners.c:11:10: fatal error: 'Python.h' file not found
      #include "Python.h"
               ^~~~~~~~~~
      1 error generated.
      error: command '/usr/bin/gcc' failed with exit code 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for biopython
ERROR: Could not build wheels for biopython, which is required to install pyproject.toml-based projects
```
These look like the same errors as before, but I am pretty confident I did it correctly.

**Comment by BradyAJohnston** [comment]
Thanks @Luc1100 for the details. Yours works otherwise by creating and linking the conda environment still?

**Comment by BradyAJohnston** [comment]
Actually @Luc1100 it seems that the error is with the `biopython` install itself.  I just had a look at the [biopython pypi page](https://pypi.org/project/biopython/1.81/#files) and it looks like there isn't a MacOS ARM build.... looks like we will have the same issue 🫠

**Comment by Luc1100** [comment]
@BradyAJohnston much sadness :( and my linking of the conda environment broke when I updated to MacOS 13 because there are some weird permissions I cannot figure out.

**Comment by mhbokhari** [comment]
Any solution to fixing the biopython install?

**Comment by BradyAJohnston** [comment]
@mhbokhari using the workaround that is currently in the documentation should still work ([docs](https://bradyajohnston.github.io/MolecularNodes/installation.html#installing-biotite-mdanalysis) as it is basically the same issue as it was previously unfortunately

**Comment by mhbokhari** [comment]
@BradyAJohnston I still get the following error message when I try out the workaround in the documentation;

note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for biopython
Failed to build biopython
ERROR: Could not build wheels for biopython, which is required to install pyproject.toml-based projects

Im using M1 with 12.5 Montrey

**Comment by BradyAJohnston** [comment]
@mhbokhari I would need more of the earlier part of the error message to help try and debug at all, the final part that you've given doesn't provide much information to work with. There are a number of other potential fixes in this thread that seem to work / not work for various people on various machines. 

I would suggest trying some of those, and if it still isn't working include more of the error messages in another comment. You can include `code` formatted text by doing three backticks and it will format it more nicely
````md
```
error message goes here
```
````

**Comment by BradyAJohnston** [comment]
The `biopython` devs are [happy to add a .whl](https://github.com/biopython/biopython/issues/4315) so this should hopefully fix another dependency issue. 

As I don't have an M1/2 mac I can't check for myself, is anyone able to try and see if any of the other dependencies are also lacking a `.whl` that we can try and get one for? Might be getting close to it finally working 🥹

**Comment by mhbokhari** [comment]
> Here is an optimized solution of MN package installation for Mac on M1/M2 Apple Silicon :
> 
> 1. Make sure your Blender is built on apple silicon version and MN has already installed.
> 2. Check if you have already install `python@3.10` via `conda` or `homebrew`: `which python3.10` or `brew list python@3.10`. If yes, go to `step 6`.
> 3. Install `miniconda` or `anaconda` on ARM version.
> 4. Create a new conda env containing python3.10: `conda create -n py310 python=3.10 -y`
> 5. Activate it: `conda activate py310`
> 6. walk to the build-in python binary dir of blender: `cd /Applications/Blender.app/Contents/Resources/3.4/python/bin`
> 7. run pip with a CPATH anoucement:
>    ```
>    CPATH=$CPATH:$(python3.10-config --include | sed  -E 's/ ?-I/:/g') ./pip3.10 install MDAnalysis==2.2.0
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    To install the exactly pinned version of required packages in `requirements.txt`, please use the following instead:
>    ```
>    CPATH=$CPATH:$(python3.10-config --include | sed  -E 's/ ?-I/:/g') ./pip3.10 install -r "/Users/$(whoami)/Library/Application Support/Blender/3.4/scripts/addons/MolecularNodes/requirements.txt"
>    ```
> 
> <img alt="image" width="1218" src="https://user-images.githubusercontent.com/33014714/218696685-3edea474-3f27-4840-bd65-6b5b9644bc74.png">

This worked for me!

**Comment by Luc1100** [comment]
Thanks @mhbokhari , just had to copy past that last line and everything is working perfectly!

**Comment by Aseamann** [comment]
Attempted to follow steps for @mhbokhari , however, pip3.10 does not seem to be in that directory with blender 3.5.1. When I move into the directory of "/Applications/Blender.app/Contents/Resources/3.5/python/bin", pip310 is not there. Only see "python3.10" in that directory. I have the conda env created and activated but CPATH step is getting an error saying "zsh: no such file or directory: ./pip3.10".

I previously had MolNodes working in a python env with conda but updated and broke what I was doing.

**Comment by Aseamann** [comment]
I didn't have to do the above step with Blender 3.5.0 and MolNodes 2.7.4. Happy to continue to help, have a bit more time this summer. Excited for the Ribbon structure update @BradyAJohnston

**Comment by BradyAJohnston** [comment]
@Aseamann does your last message mean that it is now working in Blender 3.5.0 / MN 2.7.4? Happy for any help you would like to offer. Feel free to open separate issues / PRs to discuss particular topics.

**Comment by Aseamann** [comment]
@BradyAJohnston Correct, with 3.5.0 I didn't have to mess with the python environment at all. However, I attempted with Blender 3.5.1 and it did not, I was unable to see the properties title in the "add-ons" menu under MN to install the packages.

**Comment by BradyAJohnston** [comment]
@Aseamann upgrading can be weird sometimes. I would uninstall MN from 3.5.0, uninstall and reinstall 3.5.1, then reinstall MN and see if that helps.

**Comment by YaoYinYing** [comment]
> Attempted to follow steps for @mhbokhari , however, pip3.10 does not seem to be in that directory with blender 3.5.1. When I move into the directory of "/Applications/Blender.app/Contents/Resources/3.5/python/bin", pip310 is not there. Only see "python3.10" in that directory. I have the conda env created and activated but CPATH step is getting an error saying "zsh: no such file or directory: ./pip3.10".
> 
> I previously had MolNodes working in a python env with conda but updated and broke what I was doing.

`./pip3.10` will be created only after the `./python3.10 -m ensurepip` and `./python3.10 -m pip install pip --upgrade` commands are executed.

For a detailed implementation, you may see this outdated installer script hahaha: https://github.com/YaoYinYing/MolecularNodes/blob/ARM-mac-installer/MolecularNodes/troubleshoot/InstallDependencies_M1.sh

However, please note that using this script may potentially cause issues with your Python/pip linking. So proceed with caution and make sure to handle it carefully.

**Comment by m-a-r-i-u-s** [comment]
Hi all,
I had some issues to install MolecularNodes 2.7.4 under Blender3.6.0 on a MacBook Pro M1. I found a workaround and I wanted to share my workflow here, in case somebody else has similar troubles:

1. Followed the instructions on: https://bradyajohnston.github.io/MolecularNodes/installation.html. However I added in front of "python -m pip wheel MDAnalysis==2.5.0 --cache-dir ."  the flag: ARCHFLAGS="-arch arm64".
2. This step "./python3.10 -m pip install MDAnalysis --cache-dir ~/MDAnalysis-wheel" raised an error saying it cannot install Biopython due to missing Python.h header file. So, I downloaded the missing header files from the Python3.10.12 distribution (https://www.python.org/downloads/source/). 
3. Under macOS (Ventura 13.4) I could not simply copy the header files to Blender's Python distribution. Hence I did the following:

    - cd /Applications/Blender.app/Contents/Resources/3.6/python/include
    - mv python3.10 python3.10.bk
    -  cp -r ~/Downloads/Python-3.10.12/Include . 
    - mv Include python3.10
    -  cp python3.10.bk/pyconfig.h python3.10/
    -  cd /Applications/Blender.app/Contents/Resources/3.6/python/bin
4. I installed then the missing packages with: "ARCHFLAGS="-arch arm64" ./python3.10 -m pip install MDAnalysis --cache-dir ~/Documents/mdanalysis_wheel/". The reason is that Biopython's Bio.Align will raise an error if a trajectory is loaded complaining about the wrong architecture (-86x_64 instead of arm64). 
5. MacOS's security settings will tell you that the modified Blender application is damaged and should be moved to the bin. Under Settings>Privacy & Security>Developer Tools you can add Blender to the list. I was then able to successfully start Blender from the Terminal and load a xtc trajectory.

I am happy to get your feedback about my suggestions and I hope I can help somebody!
Cheers,
Marius

**Comment by BradyAJohnston** [comment]
Hi @m-a-r-i-u-s, thanks for the details! Good to have more examples of things that work for different people. I'm still not sure why the `./python3.10 -m pip install MDAnalysis --cache-dir ~/MDAnalysis-wheel` would complain about missing headers. It should not be trying to use Blender's python, it should be using the conda environment's python which should have the required headers, but that error seems to say that it is attempting to use Blender's python.

Another solution should be just ensuring that you are definitely using the newly created python environoment. If you are definitely using the conda environment & it's installed python, but still getting the header error, then that is a problem with the conda environment.

**Comment by gbayarri** [comment]
Hi, I'm using M1 Monterey 12.4 with Blender 3.6.1 and I had the same problem. In my case, after doing the @mhbokhari workaround the problem persisted, so I installed directly MDAnalysis (not the cached `.whl`):

`./python3.10 -m pip install MDAnalysis`

And it works, I'm not sure why, and probably it's not the best practice.

**Comment by MarcoVando** [comment]
Hi
I've followed the installation guide for M1/M2 installing from pip doesn't return any errors.
I can also import MDA from the builtin python in Blender

But from the molecular node add-in in preferences the button is styll deactivated and seems that it doesn't detect the installation.
Is this a known issue?

**Comment by BradyAJohnston** [comment]
Hi @MarcoVando, it seems there are two different issues here. 

- `.cif` Files: Can you please open a new issue regarding the opening of the `.cif` files as that doesn't involve MDA. When you open the `.cif` can you try and include an example file that doesn't work and any additional details that could help with debugging.

- MDA Installation: Can you please report Blender version, Molecular Nodes version, more details about your mac, what you have tried in terms of installation, as currently there are not many details to go on.

**Comment by BradyAJohnston** [comment]
For those who have been having trouble with the installation instructions on the website, I have realised that it might be because there last step that installs from the cache doesn't specify the version to install, which  seems to sometimes install the more recent versions from pip which are higher, and thus it's not detected when installed. I will update the documentation but specifying the specific version may also help with installation issues.

#### Old Instrcutions
``` bash
./python3.10 -m pip install MDAnalysis --cache-dir ~/MDAnalysis-wheel
```

#### New Instructions
``` bash
./python3.10 -m pip install MDAnalysis==2.5.0 --cache-dir ~/MDAnalysis-wheel
```

**Comment by MarcoVando** [comment]
This solved my installation problem. Now also blender recognize MDA package as installed. 
I can also import structure form pdb databse

**Comment by BradyAJohnston** [comment]
There should now be a version that is no longer problematic to install on macOS ARM machines. If people in this thread could please try it out by downloading it from the PR #400 and try it out installing. 

If others report back that it's working, then I can move it into main and this saga can finally come to an end.

**Comment by Luc1100** [comment]
Hi @BradyAJohnston,

I was able to download MDAnalysis through the plugin page successfully. However, when I try to load a trajectory I get the following error:

Python: Traceback (most recent call last):
  File "/Users/lmorley/Library/Application Support/Blender/3.6/scripts/addons/molecularnodes/io/md.py", line 131, in execute
    mol, universe = load(
  File "/Users/lmorley/Library/Application Support/Blender/3.6/scripts/addons/molecularnodes/io/md.py", line 99, in load
    mol = session.show(atoms=universe,
  File "/Users/lmorley/Library/Application Support/Blender/3.6/scripts/addons/molecularnodes/io/mda.py", line 534, in show
    mol_object = self._process_atomgroup(
  File "/Users/lmorley/Library/Application Support/Blender/3.6/scripts/addons/molecularnodes/io/mda.py", line 774, in _process_atomgroup
    nodes.create_starting_node_tree(
  File "/Users/lmorley/Library/Application Support/Blender/3.6/scripts/addons/molecularnodes/blender/nodes.py", line 388, in create_starting_node_tree
    group = new_group(name)
  File "/Users/lmorley/Library/Application Support/Blender/3.6/scripts/addons/molecularnodes/blender/nodes.py", line 229, in new_group
    group.interface.new_socket('Geometry', in_out='INPUT', socket_type='NodeSocketGeometry')
AttributeError: 'GeometryNodeTree' object has no attribute 'interface'

**Comment by BradyAJohnston** [comment]
@Luc1100 that looks like you might be on a 3.X version of Blender, but the latest MN requires 4.0+.

**Comment by Luc1100** [comment]
@BradyAJohnston oops my bad. New error though, unsure if it something I am doing:

Python: Traceback (most recent call last):
  File "/Users/lmorley/Library/Application Support/Blender/4.0/scripts/addons/molecularnodes/io/md.py", line 131, in execute
    mol, universe = load(
  File "/Users/lmorley/Library/Application Support/Blender/4.0/scripts/addons/molecularnodes/io/md.py", line 89, in load
    universe = mda.Universe(top, traj)
TypeError: MockUniverse() takes no arguments

**Comment by BradyAJohnston** [comment]
@Luc1100 after install try restarting Blender and see if that error goes away.

**Comment by Luc1100** [comment]
@BradyAJohnston that fixed it 🙌
